src/lib/__tests__/index-test.ts

```diff
@@ -0,0 +1,855 @@
+        mainHelper,
+      });
+
+      instance.init!(
+        createInitOptions({
+          instantSearchInstance,
+        })
+      );
+
+      expect(instantSearchInstance.scheduleStalledRender).toHaveBeenCalledTimes(
+        0
+      );
+
+      // Simulate a call to search from a widget
+      instance.getHelper()!.search();
+
+      expect(instantSearchInstance.scheduleStalledRender).toHaveBeenCalledTimes(
+        1
+      );
+    });
+
+    it('calls `init` on its widgets', () => {
+      const instance = index({ indexName: 'index_name' });
+      const instantSearchInstance = createInstantSearch();
+
+      const searchBox = createSearchBox({
+        init: jest.fn(),
+      });
+
+      const pagination = createSearchBox({
+        init: jest.fn(),
+      });
+
+      instance.addWidgets([searchBox, pagination]);
+
+      expect(searchBox.init).toHaveBeenCalledTimes(0);
+      expect(pagination.init).toHaveBeenCalledTimes(0);
+
+      instance.init!(
+        createInitOptions({
+          instantSearchInstance,
+        })
+      );
+
+      expect(searchBox.init).toHaveBeenCalledTimes(1);
+      expect(pagination.init).toHaveBeenCalledTimes(1);
+      expect(pagination.init).toHaveBeenCalledWith({
+        instantSearchInstance,
+        parent: instance,
+        helper: instance.getHelper(),
+        state: instance.getHelper()!.state,
+        templatesConfig: instantSearchInstance.templatesConfig,
+        createURL: instantSearchInstance._createAbsoluteURL,
+        onHistoryChange: instantSearchInstance._onHistoryChange,
+      });
+    });
+  });
+
+  describe('render', () => {
+    it('calls `render` on its widgets', async () => {
+      const instance = index({ indexName: 'index_name' });
+      const searchClient = createSearchClient();
+      const mainHelper = algoliasearchHelper(searchClient, '', {});
+      const instantSearchInstance = createInstantSearch({
+        mainHelper,
+      });
+
+      const searchBox = createSearchBox({
+        render: jest.fn(),
+      });
+
+      const pagination = createSearchBox({
+        render: jest.fn(),
+      });
+
+      instance.addWidgets([searchBox, pagination]);
+
+      expect(searchBox.render).toHaveBeenCalledTimes(0);
+      expect(pagination.render).toHaveBeenCalledTimes(0);
+
+      instance.init!(
+        createInitOptions({
+          instantSearchInstance,
+        })
+      );
+
+      // Simulate a call to search from a widget - this step is required otherwise
+      // the DerivedHelper does not contain the results. The `lastResults` attribute
+      // is set once the `result` event is emitted.
+      instance.getHelper()!.search();
+
+      await runAllMicroTasks();
+
+      expect(searchBox.render).toHaveBeenCalledTimes(0);
+      expect(pagination.render).toHaveBeenCalledTimes(0);
+
+      instance.render!(
+        createRenderOptions({
+          instantSearchInstance,
+        })
+      );
+
+      expect(searchBox.render).toHaveBeenCalledTimes(1);
```

Comment:
We don't test what `searchBox.render` was called with?

Reply:
bd2b7f77ad8405f4b862d0929064896905ef03ad